include/swift/Syntax/RawSyntax.h

```diff
@@ -218,104 +207,225 @@ struct SyntaxPrintOptions {
 /// RawSyntax - the strictly immutable, shared backing nodes for all syntax.
 ///
 /// This is implementation detail - do not expose it in public API.
-struct RawSyntax : public llvm::ThreadSafeRefCountedBase<RawSyntax> {
-
-  using LayoutList = std::vector<RC<RawSyntax>>;
-
-  /// The kind of syntax this node represents.
-  const SyntaxKind Kind;
-
-  /// The "layout" of the node - representing the children, or the terms
-  /// in the production of the grammar.
-  const LayoutList Layout;
-
-  /// Whether this piece of syntax was actually present in the source.
-  const SourcePresence Presence;
+class RawSyntax final
+    : public llvm::ThreadSafeRefCountedBase<RawSyntax>,
+      private llvm::TrailingObjects<RawSyntax, RC<RawSyntax>, OwnedString,
+                                    TriviaPiece> {
+  friend TrailingObjects;
+
+  union {
+    uint64_t Clear;
+    struct {
+      /// The kind of syntax this node represents.
+      unsigned Kind : bitmax(NumSyntaxKindBits, 8);
+      /// Whether this piece of syntax was actually present in the source.
+      unsigned Presence : 1;
+    };
+    enum { NumRawSyntaxBits = bitmax(NumSyntaxKindBits, 8) + 1 };
+
+    // For "layout" nodes.
+    struct {
+      uint64_t : bitmax(NumRawSyntaxBits, 32);
+      /// Number of children this "layout" node has.
+      unsigned NumChildren : 32;
+    };
+
+    // For "token" nodes.
+    struct {
+      uint64_t : bitmax(NumRawSyntaxBits, 16);
+      /// The kind of token this "token" node represents.
+      tok TokenKind : 16;
+      /// Number of leading  trivia pieces.
+      unsigned NumLeadingTrivia : 16;
+      /// Number of trailing trivia pieces.
+      unsigned NumTrailingTrivia : 16;
+    };
+  } Bits;
+
+  size_t numTrailingObjects(OverloadToken<RC<RawSyntax>>) const {
+    return isToken() ? 0 : Bits.NumChildren;
+  }
+  size_t numTrailingObjects(OverloadToken<OwnedString>) const {
+    return isToken() ? 1 : 0;
+  }
+  size_t numTrailingObjects(OverloadToken<TriviaPiece>) const {
+    return isToken() ? Bits.NumLeadingTrivia + Bits.NumTrailingTrivia : 0;
+  }
```

Comment:
Not a super big deal but we shouldn't need to define `numTrailingObjects` for the last trailing object type. It will never get called.

Reply:
Yeah, but I'd rather keep this so we can see how we calculate the number of trailing objects.